WSResponse.asXml() not work with no UTF-8 encoding

I use WSClient to access a remote service. I got the following response:

Transfer-Encoding: chunked
Content-Type: application/xml;charset=euc-kr

When I use
     Document doc = response.asXml() ;
for this resposne, I got:

application - org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 81; Invalid byte 1 of 1-byte UTF-8 sequence.
...
Caused by: org.xml.sax.SAXParseException: Invalid byte 1 of 1-byte UTF-8 sequence.
        at org.apache.xerces.parsers.DOMParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.DocumentBuilderImpl.parse(Unknown Source)
        at play.libs.XML.fromInputSource(XML.java:76)
        at play.libs.XML.fromInputStream(XML.java:56)
        at play.libs.ws.ahc.AhcWSResponse.asXml(AhcWSResponse.java:127)
 
It is right that at column 81 starting the Korean character.

If I use response.getBody() to get the string representation, and then pass the string to Jaxb unmarshall, everything works as exptected.

--- a/framework/src/play-integration-test/src/test/scala/play/it/libs/WSSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/libs/WSSpec.scala
@@ -4,31 +4,28 @@
 package play.it.libs
 
 import java.io.File
+import java.nio.ByteBuffer
+import java.nio.charset.{ Charset, StandardCharsets }
 import java.util
 import java.util.concurrent.TimeUnit
 
-import akka.stream.scaladsl.FileIO
+import akka.stream.scaladsl.{ FileIO, Sink, Source }
 import akka.util.ByteString
-import akka.stream.scaladsl.Source
-import akka.stream.scaladsl.Sink
 import org.asynchttpclient.{ RequestBuilderBase, SignatureCalculator }
 import play.api.http.Port
 import play.api.libs.json.JsString
 import play.api.libs.oauth._
+import play.api.libs.streams.Accumulator
+import play.api.mvc.Results.Ok
 import play.api.mvc._
 import play.api.test._
 import play.core.server.Server
 import play.it._
 import play.it.tools.HttpBinApplication
-import play.api.mvc.Results.Ok
-import play.api.libs.streams.Accumulator
-import play.api.libs.ws.StreamedBody
-import play.libs.ws.WSResponse
 import play.mvc.Http
 
-import scala.concurrent.Await
 import scala.concurrent.duration._
-import scala.concurrent.Future
+import scala.concurrent.{ Await, Future }
 
 object NettyWSSpec extends WSSpec with NettyIntegrationSpecification
 
@@ -43,6 +40,15 @@ trait WSSpec extends PlaySpecification with ServerIntegrationSpecification {
   def app = HttpBinApplication.app
 
   val foldingSink = Sink.fold[ByteString, ByteString](ByteString.empty)((state, bs) => state ++ bs)
+
+  val isoString = {
+    // Converts the String "Hello €" to the ISO Counterparty
+    val sourceCharset = StandardCharsets.UTF_8
+    val buffer = ByteBuffer.wrap("Hello €".getBytes(sourceCharset))
+    val data = sourceCharset.decode(buffer)
+    val targetCharset = Charset.forName("Windows-1252")
+    new String(targetCharset.encode(data).array(), targetCharset)
+  }
 
   "WS@java" should {
 
@@ -92,6 +98,17 @@ trait WSSpec extends PlaySpecification with ServerIntegrationSpecification {
           val contentLength = req.headers.get(CONTENT_LENGTH)
           val transferEncoding = req.headers.get(TRANSFER_ENCODING)
           Ok(s"Content-Length: ${contentLength.getOrElse(-1)}; Transfer-Encoding: ${transferEncoding.getOrElse(-1)}")
+        }
+      } { implicit port =>
+        withClient(block)
+      }
+    }
+
+    def withXmlServer[T](block: play.libs.ws.WSClient => T) = {
+      Server.withRouter() {
+        case _ => Action { req =>
+          val elem = <name>{ isoString }</name>.toString()
+          Ok(elem).as("application/xml;charset=Windows-1252")
         }
       } { implicit port =>
         withClient(block)
@@ -198,6 +215,11 @@ trait WSSpec extends PlaySpecification with ServerIntegrationSpecification {
       body.path("file").textValue() must_== "This is a test asset."
     }
 
+    "response asXml with correct contentType" in withXmlServer { ws =>
+      val body = ws.url("/xml").get().toCompletableFuture.get().asXml()
+      new String(body.getElementsByTagName("name").item(0).getTextContent.getBytes("Windows-1252")) must_== isoString
+    }
+
     class CustomSigner extends WSSignatureCalculator with org.asynchttpclient.SignatureCalculator {
       def calculateAndAddSignature(request: org.asynchttpclient.Request, requestBuilder: org.asynchttpclient.RequestBuilderBase[_]) = {
         // do nothing
@@ -220,8 +242,7 @@ trait WSSpec extends PlaySpecification with ServerIntegrationSpecification {
 
   "WS@scala" should {
 
-    import play.api.libs.ws.WSSignatureCalculator
-    import play.api.libs.ws.StreamedBody
+    import play.api.libs.ws.{ StreamedBody, WSSignatureCalculator }
 
     implicit val materializer = app.materializer
 

--- a/framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSResponse.java
+++ b/framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSResponse.java
@@ -18,7 +18,9 @@ import java.net.URI;
 import java.net.URISyntaxException;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
-import java.util.*;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
 
 import static java.util.stream.Collectors.toList;
 
@@ -98,15 +100,20 @@ public class AhcWSResponse implements WSResponse {
         return null;
     }
 
-    public String getBody() {
-        // RFC-2616#3.7.1 states that any text/* mime type should default to ISO-8859-1 charset if not
-        // explicitly set, while Plays default encoding is UTF-8.  So, use UTF-8 if charset is not explicitly
-        // set and content type is not text/*, otherwise default to ISO-8859-1
+    private String contentType() {
         String contentType = ahcResponse.getContentType();
         if (contentType == null) {
             // As defined by RFC-2616#7.2.1
             contentType = "application/octet-stream";
         }
+        return contentType;
+    }
+
+    public String getBody() {
+        // RFC-2616#3.7.1 states that any text/* mime type should default to ISO-8859-1 charset if not
+        // explicitly set, while Plays default encoding is UTF-8.  So, use UTF-8 if charset is not explicitly
+        // set and content type is not text/*, otherwise default to ISO-8859-1
+        String contentType = contentType();
         Charset charset = HttpUtils.parseCharset(contentType);
 
         if (charset != null) {
@@ -120,15 +127,22 @@ public class AhcWSResponse implements WSResponse {
 
     /**
      * Get the response body as a {@link Document DOM document}
+     *
      * @return a DOM document
      */
     @Override
     public Document asXml() {
-        return play.libs.XML.fromInputStream(ahcResponse.getResponseBodyAsStream(), "utf-8");
+        String contentType = contentType();
+        Charset charset = HttpUtils.parseCharset(contentType);
+        if (charset == null) {
+            charset = StandardCharsets.UTF_8;
+        }
+        return play.libs.XML.fromInputStream(ahcResponse.getResponseBodyAsStream(), charset.name());
     }
 
     /**
      * Get the response body as a {@link JsonNode}
+     *
      * @return the json response
      */
     @Override
@@ -139,6 +153,7 @@ public class AhcWSResponse implements WSResponse {
 
     /**
      * Get the response body as a stream
+     *
      * @return The stream to read the response body from
      */
     @Override
@@ -148,6 +163,7 @@ public class AhcWSResponse implements WSResponse {
 
     /**
      * Get the response body as a byte array
+     *
      * @return The byte array
      */
     @Override

